### PR TITLE
transit relay: add a connection timeout of 60 seconds

### PR DESCRIPTION
**Pull request by vu3rdd**
Let us say, a user sent a file, a recipient typed in the code and then
got the transit message via wormhole and then got the offer
message. The recipient is shown the filename/size and asked for
consent. If the sender disconnects at that point or closes the
program before the content is sent, the recipient shouldn't wait
indefinitely trying to make the connection (or be at the mercy of
operating system timeouts)

The present change adds a "read/write" deadline. i.e. If Read() or
Write() takes more than the specified time (60 seconds in this
commit), then we disconnect the connection to the relay server.

Closes #OP858

---

## Code Review Checklist (to be filled out by reviewer)

- [ ] Description accurately reflects what changes are being made.
- [ ] Description explains why the changes are being made (or references an issue containing one).
- [ ] The PR appropriately sized.
- [ ] New code has enough tests. 
- [ ] New code has enough documentation to answer "how do I use it?" and "what does it do?".
- [ ] Existing documentation is up-to-date, if impacted.

### CODE REVIEW

**Review by ewanas — APPROVED** [review]
I tested killing the sender after the offer is sent, now the receiver times out instead of hanging after rejecting or accepting the offer.

One suggestion I have would be to keep all the low-level settings(message size, various timeouts, etc) in a struct and have a default instance of this struct used in absence of a caller-provided one. (Not a requirement for merging though)

**Comment by piegamesde** [comment]
It looks to me like masking symptoms instead of treating the root cause. The cause is that our general cancellation process is lacking at the moment. Adding the timeout should be the last thing to do, and I fear that otherwise we will never get to fix all the other issues.

The only scenarios where a timeout should be triggered are actual connection loss and the other side crashed without graceful shutdown.

**Comment by ewanas** [comment]
I agree that we *should* have a way to gracefully end the process for the other side. However, in case that process fails for whatever reason, or if there is connection loss, the timeouts will be handy in order to prevent hanging.

RE cancellation: I think either the file transfer protocol should have its own cancellation message or the client should respond to a close message by terminating the file transfer. I believe right now the close message is being received but isn't really terminating the client (@vu3rdd can confirm if that's the case as I haven't tested this myself)

**Comment by vu3rdd** [comment]
I guess `close()` is for mailbox. For transit: while the file transfer is in progress, if one of them cancels, at the moment, we are killing the context. So, one arm of the connection to the relay would disappear. Now the other side doesn't know about it. But relay knows about it and I think [this part of the code](https://github.com/magic-wormhole/magic-wormhole-transit-relay/blob/master/src/wormhole_transit_relay/server_state.py#L459) in the relay server takes care of disconnecting the partner and further writes/reads would error out.

**Comment by vu3rdd** [comment]
> The only scenarios where a timeout should be triggered are actual connection loss and the other side crashed without graceful shutdown.

I believe the ticket describes exactly this? Sender deliberately kills the process before the recipient even accepts the connection, so that connection to the relay has disappeared. fwiw, this is the behaviour in the python client too.

**Comment by meejah** [comment]
Semantically, `close()`-ing the mailbox should cancel any / all pending or active transfers probably (although the specs don't say one way or another).

The Python implementation does the whole transfer, then `close()`-es the mailbox at the end so that's probably what the specs _should_ say.

File: wormhole/file_transport.go [hunk 1]
```diff
@@ -269,6 +274,12 @@ func (t *fileTransport) connectToRelay(ctx context.Context, successChan chan net
...
 		conn = websocket.NetConn(ctx, wsconn, websocket.MessageBinary)
+
+		err = conn.SetDeadline(time.Now().Add(time.Second * 60))
+		if err != nil {
+			failChan <- addr
+			return
+		}
 	}
 
 	_, err = conn.Write(t.relayHandshakeHeader())
```
Comment (ewanas) [inline wormhole/file_transport.go:285]: I think after the `Read` or `Write` you should probably add a `conn.SetDeadLine(time.Time{})` in order to remove the timeout for other operations using the same connection.

https://pkg.go.dev/net#Conn.SetDeadline

**Pull request closed without merging**